### PR TITLE
Implement virtual stake delegate

### PR DESCRIPTION
**Pull request by alpe**
* Mint "virtual" tokens and delegate

### CODE REVIEW

File: x/meshsecurity/keeper/stake.go [hunk 1]
```diff
@@ -2,10 +2,66 @@ package keeper
...
+	validator, found := k.staking.GetValidator(pCtx, valAddr)
+	if !found {
+		return sdk.ZeroDec(), stakingtypes.ErrNoValidatorFound
+	}
+
+	// Ensure MS constraints:
+	newTotalDelegatedAmount := k.getTotalDelegatedAmount(pCtx, actor).Add(amt.Amount)
+	max := k.GetMaxCapLimit(pCtx, actor).Amount
+	if newTotalDelegatedAmount.GT(max) {
+		return sdk.ZeroDec(), types.ErrMaxCapExceeded
```
Comment (ethanfrey) [inline x/meshsecurity/keeper/stake.go:32]: Do you want a different error when maxcap is 0.

One is "You could do 1000, but tried 1200". The other is "You haven't been authorized at all".
Different error messages, or including the cap amount in the error message would help debugging

File: x/meshsecurity/keeper/stake.go [hunk 2]
```diff
@@ -2,10 +2,66 @@ package keeper
...
+	err = k.bank.SendCoinsFromModuleToAccount(cacheCtx, types.ModuleName, actor, coins)
+	if err != nil {
+		return sdk.ZeroDec(), err
+	}
+	// delegate virtual coins to the validator
+	newShares, err := k.staking.Delegate(
+		cacheCtx,
+		actor,
+		amt.Amount,
+		stakingtypes.Unbonded,
```
Comment (ethanfrey) [inline x/meshsecurity/keeper/stake.go:53]: Huh, why Unbonded?
(I didn't hunt down the SDK docs to see proper usage, but this stood out)

In the test case, the validator may be registered but not in the active set (which is what I think Unbonded means). Anyway, only thing that caught my eye

File: x/meshsecurity/keeper/stake.go [hunk 3]
```diff
@@ -2,10 +2,66 @@ package keeper
...
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/errors"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	types "github.com/osmosis-labs/mesh-security-sdk/x/meshsecurity/types"
 )
 
-func (k Keeper) Delegate(pCtx sdk.Context, actor sdk.AccAddress, valAddr sdk.ValAddress, amt sdk.Coin) error {
-	panic("not implemented, yet")
+// Delegate delegate mints new "virtual" bonding tokens and delegates them to the given validator.
```
Comment (ethanfrey) [inline x/meshsecurity/keeper/stake.go:11]: Function looks very solid and nice idea with the cacheWrap to make all changes atomic (although the whole call would probably be in such a CacheWrap anyway, right?

**Review by ethanfrey — APPROVED** [review]
Good stuff. Left a few questions (due to my limited knowledge of SDK), but moving well

File: x/meshsecurity/keeper/stake.go (hunk 1 above)
Comment (alpe) [reply x/meshsecurity/keeper/stake.go:32]: 👍 Adding numbers here will be helpful. Good idea!
Important but authorization should be handled before entering this method as separation of concerns. I will add a note to the doc.

File: x/meshsecurity/keeper/stake.go (hunk 2 above)
Comment (alpe) [reply x/meshsecurity/keeper/stake.go:53]: I had stumbled upon this as well but it is the same as used for any delegate msg: https://github.com/cosmos/cosmos-sdk/blob/main/x/staking/keeper/msg_server.go#L253

File: x/meshsecurity/keeper/stake.go (hunk 3 above)
Comment (alpe) [reply x/meshsecurity/keeper/stake.go:11]: When triggered by external TX then this is always in a cache wrap. Not necessarily when called by a begin/endblock.
The cache wrap should always be at a higher level to cover the full process and not the method only but I copied this from Osmosis as an extra safety net.

**Pull request merged**